### PR TITLE
Add remove action to campaign item

### PR DESCRIPTION
**Pull request by phalkunz**
Issue: https://github.com/silverstripe/silverstripe-campaign-admin/issues/39

### CODE REVIEW

**Comment by dhensby** [comment]
needs rebasing/building

File: client/src/containers/CampaignAdmin/CampaignAdmin.js [hunk 1]
```diff
@@ -206,13 +218,49 @@ class CampaignAdmin extends SilverStripeComponent {
...
+By removing this item all linked items will be removed unless used elsewhere.`;
+    const msg = i18n._t('CampaignAdmin.REMOVE_ITEM_MESSAGE', fallbackMsg);
+    // eslint-disable-next-line no-alert
+    const confirmed = window.confirm(msg);
+
+    if (!confirmed) {
+      return false;
+    }
+
+    this.props.campaignActions.removeCampaignItem(
```
Comment (flamerohr) [inline client/src/containers/CampaignAdmin/CampaignAdmin.js]: you should return this promise, rather than an empty `true`

File: client/src/containers/CampaignAdmin/CampaignAdmin.js [hunk 2]
```diff
@@ -206,13 +218,49 @@ class CampaignAdmin extends SilverStripeComponent {
...
+      this.removeCampaignItemApi,
+      campaignId,
+      itemId
+    )
+      .then(() => {
+        const endpoint = this.props.sectionConfig.readCampaignsEndpoint;
+        const fetchURL = endpoint.url;
+        this.props.recordActions.fetchRecords(
```
Comment (flamerohr) [inline client/src/containers/CampaignAdmin/CampaignAdmin.js]: you can return this inside the `then`

File: client/src/containers/CampaignAdmin/CampaignAdmin.js [hunk 3]
```diff
@@ -206,13 +218,49 @@ class CampaignAdmin extends SilverStripeComponent {
...
+          this.props.sectionConfig.treeClass,
+          endpoint.method,
+          fetchURL
+        )
+          .then(() => {
```
Comment (flamerohr) [inline client/src/containers/CampaignAdmin/CampaignAdmin.js]: after returning `this.props.recordActions.fetchRecords(` you can move this to outside the parent `then`

So you'll have:
```
return this.props.campaignActions.removeCampaignItem()
.then(return fetchRecords)
.then(selectChangeSetItem());
```

Welcome to promise chains!

File: client/src/containers/CampaignAdmin/CampaignAdmin.js [hunk 4]
```diff
@@ -206,13 +218,49 @@ class CampaignAdmin extends SilverStripeComponent {
...
+            this.props.campaignActions.selectChangeSetItem(null);
+            // Workaround to hide more actions popover
+            window.document.body.click();
+          });
+      });
+
+    return true;
```
Comment (flamerohr) [inline client/src/containers/CampaignAdmin/CampaignAdmin.js]: So the reason to return the promise, rather than this empty `true` is for unit testing...

I would like to see a unit test covering the functions that need to be called in this chain.

File: client/src/containers/CampaignAdmin/CampaignAdminList.js [hunk 5]
```diff
@@ -99,8 +100,12 @@ class CampaignAdminList extends SilverStripeComponent {
       selected = items.find(item => itemId === item.ID);
     }
 
-    if (!selected && items.length > 0) {
-      selected = items[0];
+    if (!selected) {
+      // Select the first item in the first display group
+      const groupItems = this.groupItemsForSet();
+      const groupItemsKeys = Object.keys(groupItems);
+      const group = groupItemsKeys.length > 0 ? groupItems[groupItemsKeys[0]] : null;
```
Comment (flamerohr) [inline client/src/containers/CampaignAdmin/CampaignAdminList.js]: move `groupItemsKeys[0]` to a variable please

File: client/src/containers/CampaignAdmin/CampaignAdminList.js [hunk 6]
```diff
@@ -274,13 +279,69 @@ class CampaignAdminList extends SilverStripeComponent {
...
+      selectedItem._links.referenced_by.length || 0;
+    const unremoveableInfoText = requiredByNum < 2 ?
+      i18n._t(
+        'CampaignAdmin.UNREMOVEABLE_INFO_SINGULAR',
+        'Required by %s item, and cannot be removed directly.'
+      ) :
+      i18n._t(
+        'CampaignAdmin.UNREMOVEABLE_INFO_PLURAL',
+        'Required by %s items, and cannot be removed directly.'
+      );
```
Comment (flamerohr) [inline client/src/containers/CampaignAdmin/CampaignAdminList.js]: I strongly don't agree with having plurals hardcoded like this... we've talked about this before as well. This doesn't translate well for languages with different pluralisation rules to english (like Russian) and duplication work for no pluralisation rules (like Chinese).

Please change this to `'Required by {number} item(s), and cannot be removed directly.'`

If you'd like we can work on (or raise a story for) a pluralisation API for the javascript side, which can be similar to what was implemented on the PHP side.

Missing entry for new translation strings in `en.js` and `en.json`

File: src/CampaignAdmin.php [hunk 7]
```diff
@@ -396,6 +406,42 @@ public function readCampaign(HTTPRequest $request)
...
+            return new HTTPResponse(null, 400);
+        }
+
+        $campaignID = $request->param('CampaignID');
+        $itemID = $request->param('ItemID');
+
+        if (!$campaignID ||
+            !is_numeric($campaignID) ||
+            !$itemID ||
+            !is_numeric($itemID)) {
```
Comment (flamerohr) [inline src/CampaignAdmin.php]: having `!$campaignID` and `!$itemID` seems redundant, if an invalid numeric ID is provided (`0`) it should return a `404` error

File: client/src/containers/CampaignAdmin/CampaignAdminList.js [hunk 8]
```diff
@@ -274,13 +279,69 @@ class CampaignAdminList extends SilverStripeComponent {
...
+          className="btn btn-secondary action"
+          onClick={this.handleRemoveItem}
+        >
+          Remove
+        </button>
+      ) :
+      (
+        <p key="unremoveable_info" className="alert alert-info campaign-admin__unremoveable-item">
+          <span className="font-icon-link"></span>
+          {i18n.sprintf(unremoveableInfoText, requiredByNum)}
```
Comment (flamerohr) [inline client/src/containers/CampaignAdmin/CampaignAdminList.js]: Please use i18n.inject() with named attributes instead of i18n.sprintf(), so we don't have positional issues later on :)

File: client/src/containers/CampaignAdmin/CampaignAdmin.js [hunk 9]
```diff
@@ -206,13 +218,49 @@ class CampaignAdmin extends SilverStripeComponent {
...
     return (
       <CampaignAdminList {...props} />
     );
   }
 
+  handleRemoveCampaignItem(campaignId, itemId) {
+    const fallbackMsg = `Are you sure you want to remove this item?
+
+By removing this item all linked items will be removed unless used elsewhere.`;
+    const msg = i18n._t('CampaignAdmin.REMOVE_ITEM_MESSAGE', fallbackMsg);
```
Comment (flamerohr) [inline client/src/containers/CampaignAdmin/CampaignAdmin.js]: Missing `en.js` and `en.json` entry for translation strings

File: client/src/containers/CampaignAdmin/CampaignAdminList.js (hunk 6 above)
Comment (phalkunz) [reply client/src/containers/CampaignAdmin/CampaignAdminList.js]: Sorry, I totally forgot about this. I've created an issue for the API (https://github.com/silverstripe/silverstripe-admin/issues/218).

File: client/src/state/campaign/tests/CampaignActions-test.js [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+describe('CampaignActions', () => {
+  const middlewares = [thunk];
+  const mockStore = configureMockStore(middlewares);
+
+  describe('removeCampaignItems()', () => {
+    it('should create 2 actions, request and success, when resolved', () => {
+      const store = mockStore({});
+      const removeItemApi = () => new Promise((resolve) => {
+        resolve();
+      });
```
Comment (flamerohr) [inline client/src/state/campaign/tests/CampaignActions-test.js]: you can shorten this to `const removeItemApi = () => Promise.resolve();` :)

**Comment by flamerohr** [comment]
Looks good, can be merged on green
